### PR TITLE
Enable Java function as alpha

### PR DESCRIPTION
**Pull request by dbadura**
**Description**

Changes proposed in this pull request:

- Enable support for Java Runtime

**Related issue(s)**
https://github.com/kyma-project/kyma/issues/15660

### CODE REVIEW

File: components/function-controller/cmd/webhook/main.go [hunk 1]
```diff
@@ -26,11 +26,12 @@ import (
 )
 
 type config struct {
-	SystemNamespace    string `envconfig:"default=kyma-system"`
-	WebhookServiceName string `envconfig:"default=serverless-webhook"`
-	WebhookSecretName  string `envconfig:"default=serverless-webhook"`
-	WebhookPort        int    `envconfig:"default=8443"`
-	WebhookConfigPath  string `envconfig:"default=/appdata/config.yaml"`
+	SystemNamespace     string               `envconfig:"default=kyma-system"`
```
Comment (dbadura) [inline components/function-controller/cmd/webhook/main.go:29]: I can move it all to package `webhook` and init config with prefix `WEBHOOK`

File: components/function-controller/internal/webhook/validating_webhook.go [hunk 2]
```diff
@@ -16,17 +16,19 @@ import (
...
-func NewValidatingHook(configv1alpha1 *serverlessv1alpha1.ValidationConfig, configv1alpha2 *serverlessv1alpha2.ValidationConfig, client ctrlclient.Client) *ValidatingWebHook {
+func NewValidatingWebHook(configv1alpha1 *serverlessv1alpha1.ValidationConfig, configv1alpha2 *serverlessv1alpha2.ValidationConfig, featureFlags FeatureFlags, client ctrlclient.Client) *ValidatingWebHook {
 	return &ValidatingWebHook{
-		configv1alpha1: configv1alpha1,
-		configv1alpha2: configv1alpha2,
-		client:         client,
+		configv1alpha1:       configv1alpha1,
+		configv1alpha2:       configv1alpha2,
+		client:               client,
+		rejectJava17AlphaJVM: !featureFlags.Java17JvmAlphaEnabled,
```
Comment (moelsayed) [inline components/function-controller/internal/webhook/validating_webhook.go:31]: Why are we using a negated field, and not the feature flag value directly?

File: components/function-controller/internal/webhook/validating_webhook.go [hunk 3]
```diff
@@ -65,9 +67,15 @@ func (w *ValidatingWebHook) handleFunctionValidation(req admission.Request) admi
 			if err := w.decoder.Decode(req, fn); err != nil {
 				return admission.Errored(http.StatusBadRequest, err)
 			}
+
+			if w.rejectJava17AlphaJVM && fn.Spec.Runtime == serverlessv1alpha2.Java17JvmAlpha {
```
Comment (moelsayed) [inline components/function-controller/internal/webhook/validating_webhook.go:71]: Nit: I think it's easier to read if we check the runtime first, then check it's enabled/rejected.

File: resources/serverless/charts/webhook/values.yaml [hunk 4]
```diff
@@ -95,6 +95,8 @@ initContainer:
     readOnlyRootFilesystem: true # Mandatory
 
 values:
+  featureFlags:
+    java17AlphaEnabled: false
```
Comment (moelsayed) [inline resources/serverless/charts/webhook/values.yaml:99]: Do you think it makes sense to move `featureFlags` to it's own top level section? I know it's technically webhook configuration, but I feel it's important enough to have it's own top level section.

File: resources/serverless/templates/runtimes.yaml [hunk 5]
```diff
@@ -73,3 +73,34 @@ data:
     RUN rm -rf /etc/pip.conf
 
     USER 1000
+---
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: dockerfile-java17-jvm-alpha
+  namespace: kyma-system
```
Comment (moelsayed) [inline resources/serverless/templates/runtimes.yaml:81]: The namespace should be configurable.

File: resources/serverless/templates/runtimes.yaml [hunk 6]
```diff
@@ -73,3 +73,34 @@ data:
...
+  labels:
+    serverless.kyma-project.io/config: runtime
+    serverless.kyma-project.io/runtime: java17-jvm-alpha
+data:
+  Dockerfile: |-
+    FROM {{ include "imageurl" (dict "reg" .Values.global.containerRegistry "img" .Values.global.images.function_runtime_java17_jvm_alpha) }} as builder
+
+    ARG BUILD_DIR=/build
+    ARG SOURCE_DIR=/src
+    ARG DEPS_DIR=/src
```
Comment (moelsayed) [inline resources/serverless/templates/runtimes.yaml:91]: Do we need both? The have the same value are used once each.

File: resources/serverless/templates/runtimes.yaml [hunk 7]
```diff
@@ -73,3 +73,34 @@ data:
...
+    WORKDIR $BUILD_DIR
+
+    COPY $DEPS_DIR/pom.xml $BUILD_DIR/handler-pom.xml
+
+    COPY $SOURCE_DIR/Handler.java $BUILD_DIR/src/main/java/io/project/kyma/serverless/handler/Handler.java
+    RUN mvn dependency:go-offline -f handler-pom.xml
+
+    RUN mvn clean && mvn package -f pom.xml
+
+    FROM eclipse-temurin:17-jre-alpine
```
Comment (moelsayed) [inline resources/serverless/templates/runtimes.yaml:101]: Should we host this image as well?

**Review by moelsayed — COMMENTED** [review]
Looks great, thank you! Just a few questions and nit comments.

File: components/function-controller/cmd/webhook/main.go (hunk 1 above)
Comment (moelsayed) [reply components/function-controller/cmd/webhook/main.go:29]: That sounds like a great idea :+1:

File: resources/serverless/templates/runtimes.yaml (hunk 6 above)
Comment (dbadura) [reply resources/serverless/templates/runtimes.yaml:91]: Unfortunately yes. I've done it to be able to use the same image for local build and kaniko build.

File: resources/serverless/templates/runtimes.yaml (hunk 7 above)
Comment (dbadura) [reply resources/serverless/templates/runtimes.yaml:101]: Yes, I will add in another PR.

File: resources/serverless/charts/webhook/values.yaml (hunk 4 above)
Comment (dbadura) [reply resources/serverless/charts/webhook/values.yaml:99]: I have a feeling that our values are distribution in the whole `values.yaml`. 
I think that it should be refactored in the future.

File: components/function-controller/internal/webhook/validating_webhook.go (hunk 2 above)
Comment (dbadura) [reply components/function-controller/internal/webhook/validating_webhook.go:31]: In context of webhook, the runtime can be rejected. 
I've done it for readability.

File: resources/serverless/templates/runtimes.yaml (hunk 6 above)
Comment (moelsayed) [reply resources/serverless/templates/runtimes.yaml:91]: Probably a good idea to add a comment about that to avoid "fixing it" in the future ;)

**Review by moelsayed — APPROVED** [review]

File: resources/serverless/templates/runtimes.yaml (hunk 6 above)
Comment (dbadura) [reply resources/serverless/templates/runtimes.yaml:92]: Added :)

**Review by moelsayed — APPROVED** [review]
LGTM

**Pull request merged**